File: iosgl/iosgl/bt/base/Camera3D.m
//
//  Camera3D.m
//  iosgl
//
//  Created by zhao on 30/12/2019.
//  Copyright © 2019 zhao. All rights reserved.
//

#import "Camera3D.h"
#import "Matrix3D.h"

@implementation Camera3D
- (instancetype)init
{
    self = [super init];
    if (self) {
        self.camMatrix3D=[[Matrix3D alloc]init];
        self.viewMatrix =[[Matrix3D alloc]init];
        self.modelMatrix =[[Matrix3D alloc]init];
        self.distance=500;
        self.sceneViewHW=100;
        self.fovw=300;
        self.fovh=500;
        [self upFrame];
    }
    return self;
}
-(void) upFrame  ;
{
  
    [self.viewMatrix identity];
//    [self.viewMatrix perspectiveFieldOfViewLH:1 aspectRatio:1 zNear:0.01 zFar:1000];
//    [self.viewMatrix appendScale:  (self.sceneViewHW / self.fovw * 2) y:self.fovw / self.fovh * ( self.sceneViewHW / self.fovw * 2) z:1];
    [self.viewMatrix inputStrData:@"1.8304877281188965,0,0,0,0,1.8304877281188965,0,0,0,0,1.0016694068908691,1,0,0,-1.0016694068908691,0"];
   
    [self.camMatrix3D identity];
    [self.camMatrix3D appendRotation: self.rotationY axis:Vector3D.Y_AXIS];
    [self.camMatrix3D appendRotation: self.rotationX axis:Vector3D.X_AXIS];
    [self.camMatrix3D appendTranslation: 0.0 y:0.0 z:self.distance];
    [self.camMatrix3D appendTranslation: 0.0 y:-0.0 z:0.0];
    
    Matrix3D* m=[self.camMatrix3D Invert];
    Vector3D* p=  [m transformVector: [[Vector3D alloc]x:0 y:0 z:-self.distance] ];
    self.x=p.x;
    self.y=p.y;
    self.z=p.z;
    self.modelMatrix= [self.viewMatrix  clone];
    [self.modelMatrix prepend:  self.camMatrix3D];
    
    
  
 
 
}
 
@end

File: h5gl/new/src/pan3d/base/Camera3D.ts
module Pan3d {
    export class Camera3D extends Object3D {
        public camMatrix3D:Matrix3D ;
        public viewMatrix:Matrix3D ;
        public modelMatrix:Matrix3D ;
        public distance:number ;
        public sceneViewHW:number ;
        public fovw:number ;
        public fovh:number ;
      
        public get cameraMatrix(): Matrix3D {
            return this.modelMatrix;
        }
        public constructor()
        {
            super()

            this.camMatrix3D=new Matrix3D();
            this.viewMatrix =new Matrix3D();
            this.modelMatrix =new Matrix3D();
            this.distance=100;
            this.sceneViewHW=100;
            this.fovw=300;
            this.fovh=500;
            this.rotationX =-30;
            this.rotationY=45;

        }
        public  upFrame  ():void
        {
    
            this.viewMatrix.identity();
            this.viewMatrix.perspectiveFieldOfViewLH(1,1,10,5000);
            this.camMatrix3D.identity();
            this.camMatrix3D.appendRotation(this.rotationY, Vector3D.Y_AXIS);
            this.camMatrix3D.appendRotation(this.rotationX, Vector3D.X_AXIS);
            this.camMatrix3D.appendTranslation(0,0,this.distance);
            this.modelMatrix= this.viewMatrix.clone();
            this.modelMatrix.prepend(this.camMatrix3D);
            var m:Matrix3D=this.camMatrix3D.clone();
            m.invert();
            var p:Vector3D=   m.transformVector(new Vector3D(0,0,-this.distance));
            this.x=p.x;
            this.y=p.y;
            this.z=p.z;
        }
    }
}